src/Routing.purs

```diff
@@ -66,24 +64,24 @@ matches
    . Match a
   -> (Maybe a -> a -> Eff (RoutingEffects eff) Unit)
   -> Eff (RoutingEffects eff) (Eff (RoutingEffects eff) Unit)
-matches = matchesWith decodeURIComponent
+matches matcher = matchesWith (matchWith decodeURIComponent matcher)
 
--- | Runs the callback on every hash change using a given custom String decoder
--- | and a `Match` parser to extract a route from the hash. If a hash fails to
--- | parse, it is ignored. To avoid dropping hashes, provide a fallback
--- | alternative in your parser. Returns an effect which will remove the
--- | listener.
+-- | Runs the callback on every hash change using a given custom parser to
+-- | extract a route from the hash. If a hash fails to parse, it is ignored.
+-- | To avoid dropping hashes, provide a fallback alternative in your parser.
+-- | Returns an effect which will remove the listener.
 matchesWith
-  :: forall eff a
-   . (String -> String)
-  -> Match a
+  :: forall eff f a
+   . Foldable f
+  => (String -> f a)
   -> (Maybe a -> a -> Eff (RoutingEffects eff) Unit)
   -> Eff (RoutingEffects eff) (Eff (RoutingEffects eff) Unit)
-matchesWith decoder matcher cb = do
-  let parser = matchWith decoder matcher
-  foldHashes
-    (\old -> either (\_ -> pure old) (\a -> cb old a $> Just a) <<< parser)
-    (pure <<< either (const Nothing) Just <<< parser)
+matchesWith parser cb = foldHashes go (go Nothing)
+  where
+  go a =
+    maybe (pure a) (\b -> Just b <$ cb a b)
```

Comment:
I was confused by this line for a while but then noticed that  `<$` is used not `<*` :d